### PR TITLE
Upgraded to esprima-fb and added optional config.esprima hook

### PR DESCRIPTION
**Pull request by trabus**
This PR changes the default esprima to esprima-fb, and adds the ability to pass a different esprima through the config. For example:

```js
// config.js
module.exports = {
  esprima: require('esprima'),
  //additional options below
}
```

### CODE REVIEW

**Comment by trabus** [comment]
I'm working on an additional test for using a different esprima through config now.

**Comment by thlorenz** [comment]
Tests failing on travis due to `^`, but feel free to remove support for `0.6-0.8`.
Rather add `0.12` and `iojs` to `.travisrc`.

**Comment by trabus** [comment]
Will do.

**Comment by thlorenz** [comment]
LGTM so far. Those extra tests would be very nice.
At least a decent start at testing ES6 features. We can extend it as needed in the future i.e. if bugs are reported.

**Comment by trabus** [comment]
Okay, done. Regarding tests, I'm not quite sure how to best test the es6 stuff. So far I've added a keyword check for `imports` and `exports`, but I could use some guidance on what other ways I should be testing.

**Comment by thlorenz** [comment]
Thanks, the other thing that'd be useful is having a default config like [this one](https://github.com/thlorenz/redeyed/blob/master/config.js) that also includes ES6 tokens. It would serve as a starting point for people providing their config.

If I remember correctly I pulled the ES5 tokens from the esprima source somehow so I'd expect you could find the ES6 tokens in the esprima-fb source as well.

File: config.js [hunk 1]
```diff
@@ -41,12 +41,14 @@ module.exports = {
...
 
     , 'else'        :  undefined
+    , 'export'      :  undefined
 
     , 'finally'     :  undefined
     , 'for'         :  undefined
     , 'function'    :  undefined
 
     , 'if'          :  undefined
+    , 'import'      :  undefined
```
Comment (thlorenz) [inline config.js]: So we should copy out this original `config.js` into `config-es5.js` and extend it afterwards so it includes **all** possible ES6 tokens.

**Comment by thlorenz** [comment]
Looked through the esprima-fb code real quick and the only thing I could find giving you token info are the tests, i.e. inside harmony features [here listing the import statement](https://github.com/facebook/esprima-fb/blob/fb-harmony/test/harmonytest.js#L5453).

It also seems that [jsx is supported in that version](https://github.com/facebook/esprima-fb/blob/fb-harmony/test/fbtest.js)l.

We could try to convince the author(s) of this  to give us a separate structure info for JSX. I wouldn't include that in the default config, but since redeyed now would support it we should provide it somewhere, i.e inside a `config-jsx.js` which users could combine with the default `config.js` (which already would include ES6 tokens).

Hopefully @facebook is able to supply the info about ES6 and JSX token structure to us easily, otherwise we'd have to dig into the tests here as well in order to pull that structure out.

cc @zpao @jeffmo

File: config.js (hunk 1 above)
Comment (trabus) [inline config.js]: :+1:

**Comment by jeffmo** [comment]
If I'm understanding correctly, you're having a hard time getting the parser to parse 'import's and 'exports'?

You can indicate "module" parsing mode by passing the `sourceType = "module"` option. This tells the parser to parse the source in "module" mode (modules and scripts parse slightly differently). Modules are implicitly strict-mode and may have `import/exports` at their toplevel; Scripts are non-explicitly strict-mode  and may not have `import`/`export`s.

As for JSX: Token info should be available if you pass the `tokens=true` option (this should JustWork™)

**Comment by trabus** [comment]
@jeffmo I'm just upgrading the esprima used to esprima-fb to provide ES6 support in the next version of redeyed. So far I've had no problem with it parsing the new keywords, I just need to add the new tokens to the config.js in redeyed.

**Comment by thlorenz** [comment]
@jeffmo basically we need to create a config like [this](https://github.com/thlorenz/redeyed/blob/master/config.js) that includes all ES6 tokens as well.

**Comment by thlorenz** [comment]
@trabus were you able to cover all ES6 keywords in the new `config.js` or do we still need help?
The jsx part is less important IMO so we could add this in another PR if it's too much of a problem to get that right.

**Comment by trabus** [comment]
@thlorenz I got all the ones [listed in the spec](http://people.mozilla.org/~jorendorff/es6-draft.html#sec-keywords), though I need to add `static` (and a test for it with "use strict"), I just noticed that I missed the note about it below the list.

I also added `=>` and `...` to punctuators, those were the only new ones I noticed [here](http://people.mozilla.org/~jorendorff/es6-draft.html#sec-punctuators).

**Comment by trabus** [comment]
Looking further, I need to add "implements", "interface", "package", "private", "protected", and "public", as they [are treated as keywords in strict mode](http://people.mozilla.org/~jorendorff/es6-draft.html#sec-identifiers-static-semantics-early-errors).

**Comment by trabus** [comment]
Scratch that, they're all [future reserved words](http://people.mozilla.org/~jorendorff/es6-draft.html#sec-future-reserved-words) (along with `enum` and `await`).

**Comment by thlorenz** [comment]
Awesome so now we have all keywords in there (except jsx ones)?
If so please squash this into one or two commits (however you see fit) and I'll merge.

File: config.js [hunk 2]
```diff
@@ -111,6 +120,10 @@ module.exports = {
     , '|=': undefined
     , '^=': undefined
     , '/=': undefined
+    
+    , '=>': undefined
+
```
Comment (thlorenz) [inline config.js]: Please remove extra empty lines here.

**Comment by trabus** [comment]
> Awesome so now we have all keywords in there (except jsx ones)?

Yes, I think I got them all.

> If so please squash this into one or two commits (however you see fit) and I'll merge.

Done!

**Comment by thlorenz** [comment]
Published as `0.5`, thanks.

**Pull request closed without merging**